Distinguish host from world on ingress

joestringer opened the pull request:
Previously, all traffic that is masqueraded to the local host's IP would be considered as coming "from the host" for policy purposes, even if the traffic originally came from outside the node. This PR introduces a magic mark on packets sourced from the local stack to ensure that local traffic is considered as "host", while traffic from the outside world (or even other containers on the host) are considered as "world".

Related: #3940 

```release-note
Distinguish host from world on ingress.
Traffic from the outside world that would previously pass through an SNAT step would have
appeared (eg, in monitor output) as though it came from `reserved:host` (`host` entity).
It is now considered `reserved:world` (`world` entity), while only traffic that comes from 
a local process will be considered as `reserved:host`.
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3988)
<!-- Reviewable:end -->

joestringer committed:
bpf: Mark traffic from outside local host as world

Pick up on the magic mark for traffic sourced from a local application,
and treat such traffic as from the host, then default to treating
traffic without this mark (or either proxy mark) as from the world.

Fixes: #3940

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Install rules to mark local applications

Add an IPTables rule to mark traffic in the OUTPUT chain (ie, traffic
coming from applications on the local host) with a magic marker, so that
when the traffic arrives in the BPF datapath we can differentiate
traffic sourced from the local node (with the node's IP) from traffic
sourced outside the local node (which was masqueraded to the node's IP).

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Reuse proxy magic marker variables

Reuse these variables from one location in golang code.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on daemon/daemon.go:
```diff
@@ -581,6 +588,20 @@ func (d *Daemon) installIptablesRules() error {
 		return err
 	}
 
+	// Mark all packets sourced from the host with a special marker so that
```
Please be more specific here and state that we mark packets from *processes* running on the host.

rlenglet commented on daemon/daemon.go:
```diff
@@ -537,12 +544,13 @@ func (d *Daemon) installIptablesRules() error {
 	// performed by kube-proxy for all packets destined for Cilium. Cilium
 	// installs a dedicated rule which does the source PAT to the right
 	// source IP.
+	clearMasqBit := fmt.Sprintf("%#04x/%#04x", 0, proxy.MagicMarkK8sMasq)
```
The option to `--set-mark` is 32-bit, and is dumped as such in iptables output. So I'd replace `%#04x` with `%#08x` to be consistent with that format.

Same for all the others below.

joestringer replied:
Added an extra commit to clean this up.

rlenglet reviewed (approved):
The proxy bits LGTM.
Just one nitpick about the formatting of marks / masks.

joestringer committed:
daemon: Format packet marks as 32bit hexits

The option to `--set-mark` is 32-bit, so it's more explicit to use
8-hexit output when formatting the commands.

Signed-off-by: Joe Stringer <joe@covalent.io>

borkmann reviewed (approved):
BPF bits LGTM
